### PR TITLE
Fix read/write microbenchmark and added production implementation to benchmark

### PR DESCRIPTION
**Pull request by m7913d**
RandomFloats(Strs)::SetUp: setlocale returns the currently set locale. To restore the old locale after the benchmark has run, the current locale should be first queried (by passinga nullptr as the second argument).
RandomFloats(Strs)::SetUp: abort the test that depend on the de_DE locale if this locale is not available to avoid producing irrelevant test results.

RandomFloatStrs: should always generate strings with a "." as the decimal separator, as xlsx files always use this as well, regardless of the current locale.

number_serialiser_mk2: `std::use_facet<std::numpunct<char>>(std::locale{}).decimal_point() == ','` always returns false, as it checks the decimal point of the "C" locale.

number_serialiser_production: uses the serialiser that is currently used by the xlnt library. Usefull to compare the currently used implementation with possible alternatives and/or monitor regression.

Enable `std::from_chars` / `std::to_chars` on all platforms if C++17 is supported.

Enable benchmarks using comma as decimal separator on all platforms. Why was this only enabled on MSVC?

Added (micro)benchmarking of release builds to CircleCI to allow monitoring regression.

### CODE REVIEW

**Comment by doomlaur** [comment]
Very nice, thank you! :smile: 

I want to take a moment and shortly talk about the following check that you added to our code:
```cpp
#if defined(_MSC_VER) || __cplusplus >= 201703L
```

First of all, from my side, it is fine for me to merge this. However, we might want to add some better checks for the future. In some of my projects, I tried to add something like this for proper C++ version checks (also on MSVC):
```cpp
/// Unfortunately, the macro __cplusplus does not report the correct version under Visual Studio unless /Zc:__cplusplus is used during compilation.
/// Source: https://learn.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-170
/// In order to have proper feature testing for C++ features that don't have feature test macros, but also to avoid forcing others
/// to configure their compiler properly, we'll need to define our own macro.
#ifndef XLNT_CPP_VERSION
#	if defined(_MSVC_LANG) && _MSVC_LANG > __cplusplus // take the larger one of the two (Microsoft does the same in the MSVC STL)
#		define XLNT_CPP_VERSION _MSVC_LANG
#	else
#		define XLNT_CPP_VERSION __cplusplus
#	endif
#endif
```

Unfortunately ugly, but for XLNT as a library, it is not be possible (to my knowledge) to do proper C++ version checks otherwise. The downside of this solution is that the file containing these defines needs to be included everywhere where C++ version checks are necessary - otherwise they obviously won't work correctly. This can be a source of unexpected issues, where the wrong code path might be used just because someone forgot to include a file.

There is, however, an even better solution to this problem - the [C++ feature test macros](https://en.cppreference.com/w/cpp/feature_test). In our specific case, it would be even better if we checked for `__cpp_lib_to_chars >= 201611L` instead of our current check. This improved check would always work on MSVC correctly, and would even enable the test for compilers that have implemented `std::to_chars` / `std::from_chars`, but do not have full C++17 support. Unfortunately, the feature test macros have been added to C++20, and although they contain macros for testing features even back to C++98, we cannot rely on them when using older compilers that do not support C++20 (and I'm not sure if even newer compilers define them anyway in C++11, C++14 or C++17 modes, although they have been officially only added in C++20).

In other words, the check you have just added will work for both newer MSVC compilers and other C++17-conformant compilers, but in such cases, we might want to have better overall version checks. Since this is not the main reason why this PR has been created, and since the check is good enough (and better than before), I'm going to approve this PR as it is. But feel free to further discuss how we should do proper version checks in the future :wink:

**Comment by doomlaur** [comment]
One more thing: we should prefer using `std::locale::classic()` rather than `std::locale("C")`, since the constructor taking a string literal will always have to search for the specified locale, and that is expensive, which is an issue in performance-critical code. Here it won't make a difference, as it is only called once during construction, but it's nevertheless a good practice :slightly_smiling_face: This is the reason why I created a function for caching the system locale in https://github.com/xlnt-community/xlnt/pull/38.

And yes, I know this piece of code calling `std::locale("C")` already existed, but I just wanted to mention it for the future :wink:

**Comment by doomlaur** [comment]
> number_serialiser_mk2: `std::use_facet<std::numpunct<char>>(std::locale{}).decimal_point() == ','` always returns false, as it checks the decimal point of the "C" locale.

Even worse: creating a [default-constructed `std::locale`](https://en.cppreference.com/w/cpp/locale/locale/locale) actually creates a copy of the `std::locale::global`, meaning that it might be the `std::locale::classic()` if none has been set, or some other locale otherwise. And this locale [will be synchronized with the C locale](https://en.cppreference.com/w/cpp/locale/locale/global) (as if `std::setlocale` has been called), unless it has no name. Unfortunately very error-prone :confused: 

In other words, it's very good that you changed it :+1:

**Review by doomlaur — DISMISSED** [review]
Thanks for this PR! :smile:

**Comment by m7913d** [comment]
I can add something like XLNT_CPP_VERSION to this PR. If we use a macro, e.g. XLNT_SUPPORT_CPP(version) it won't compile in this macro is not defined.

Using a c++20 feature to check for c++17 feature seems useless, but I'd desired we can use are own feature checks, e.g. XLNT_HAS_FEATURE(feature) which can use the c++ version in addition to compiler specific checks were needed.

**Comment by doomlaur** [comment]
> I can add something like XLNT_CPP_VERSION to this PR. If we use a macro, e.g. XLNT_SUPPORT_CPP(version) it won't compile if this macro is not defined.

Awesome idea, thank you! :+1: :smile: 

> Using a c++20 feature to check for c++17 feature seems useless

My point is that, in some cases, compiler developers might choose to backport certain newer C++ features to older versions of C++, e.g. a C++20 feature even in C++14 or C++17 modes, as long as it is beneficial and does not affect existing code in any way. Examples may be defect reports or certain performance improvements that do not affect external code, and I also think that feature test macros are something that can be easily backported without consequences. However, whether they are actually backported or not then depends on what the compiler developers decide. But, in the end, by not using such C++20 features we make sure that XLNT can also work on older compilers without issues.

The only disadvantage of the current solution is that, for example, `XLNT_CPP_VERSION` will only have a value >= `201703L` if it implements ALL the C++17 features. In other words, `XLNT_DETAIL_FEATURE_TO_CHARS` will have the incorrect value of -1 for compilers already implementing `std::to_chars` and `std::from_chars` but not yet implementing all C++17 features. But since the feature test macros have only been added in C++20, the solution you added to XLNT is the best one we can have at the moment :+1: 

> but if desired we can use are own feature checks, e.g. XLNT_HAS_FEATURE(feature) which can use the c++ version in addition to compiler specific checks were needed.

The only missing definition is `#define XLNT_CPP_98 199711L`, but since the minimum version of C++ XLNT supports is C++11, we should never need it. I'll leave it up to you if you want to add it, but I think we can leave it as it is.

Just for documentation purposes, while XLNT mostly uses C++ features, it also uses some C features. Since XLNT does not use newer C features, we don't necessarily need the version checks (at least not yet), but some of the time functions added in PR https://github.com/xlnt-community/xlnt/pull/27 (where I used some compiler-specific macros) are now part of C23 (see [localtime_r](https://en.cppreference.com/w/c/chrono/localtime) and [gmtime_r](https://en.cppreference.com/w/c/chrono/gmtime)). For C23 compilers, we could enable the usage of `localtime_r` and `gmtime_r` even for compilers that don't have the other feature test macros defined, but do support C23. Since MSVC does not have any weird rules for the C version (defined in the `__STDC_VERSION__` macro), it is easier to do and less error-prone than the C++ checks I proposed and you implemented as part of this PR, but on the other hand, defining the specific versions for each version of C/C++ (like you did in your latest commit) makes it easier to write it correctly. I'll leave it up to you whether you want to add these to this PR. Last but not least, I want to mention that MSVC [still defaults to C89 with a few C99 extensions](https://learn.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version?view=msvc-170#c-standards-support-1) by default, and only newer versions of MSVC support C11 and C17.

The following would add C version checks, in case we want to have them:
```cpp
#define XLNT_C_95 199409L
#define XLNT_C_99 199901L
#define XLNT_C_11 201112L
#define XLNT_C_17 201710L
#define XLNT_C_23 202311L
```

**Comment by doomlaur** [comment]
Last but not least, there's a merge conflict in `.circleci/continue_config_in.yml` you should solve before I approve this PR, as GitHub will probably complain otherwise :slightly_smiling_face:

**Review by doomlaur — APPROVED** [review]
Perfect, thank you! :smile: Approved!

**Comment by m7913d** [comment]
> In other words, XLNT_DETAIL_FEATURE_TO_CHARS will have the incorrect value of -1 for compilers already implementing std::to_chars and std::from_chars but not yet implementing all C++17 features. 

We can always improve the logic by using compiler specific defines and/or backported feature test macros as long as it compiles correctly for all (relevant) compiler under C++11.

> The only missing definition is #define XLNT_CPP_98 199711L

I deliberately did not include this one given we require C++11. So I would not encourage anyone to check if C++98 is supported in a PR. If we require a higher C++ version in the future (which is definitely not my intention at the moment), I would prefer to remove the C++11 version checks too.

> The following would add C version checks, in case we want to have them:

I have added the relevant C version checks (see latest commit).

**Comment by doomlaur** [comment]
Just wondering, is there any specific reason why you updated Google Benchmark from 1.5.0 to 1.8.5 instead of 1.9.0? (Not counting 1.9.1, as it has been released last week.) It's not important that we're on the very latest version, but I was just wondering whether you encountered any issues with the newer versions (which would be good to know for the future).

**Comment by m7913d** [comment]
@doomlaur I think I chose 1.8.5 instead of 1.9.0 because sometimes I prefer to avoid the *.0 releases just in case some bug is introduced by the new features.
